tests/helpers/deviceappproxy/apduabstract.py

```diff
@@ -0,0 +1,129 @@
+from typing import List, Dict, Optional, cast, NewType, Tuple, AnyStr
+from dataclasses import dataclass, field
+
+
+@dataclass
+class CApdu:
+    class Type:
+        IN: int = 0
+        OUT: int = 1
+        INOUT: int = 2
+    data: List[bytes]
+    cla: AnyStr = field(default="00")
+    ins: AnyStr = field(default="00")
+    p1: AnyStr = field(default="00")
+    p2: AnyStr = field(default="00")
+    lc: AnyStr = field(default="00")
+    le: AnyStr = field(default="00")
+    typ: Type = field(default=Type.INOUT)
+
+
+ApduDict = NewType("ApduDict", Dict[str, CApdu])
+
+
+class ApduSet:
+    """Collects a set of CApdu objects and provides their raw version, ready
+    to be sent to a device/an app"
+    """
+    _apdus: ApduDict = {}
+
+    def __init__(self, apdus: Optional[ApduDict] = None, max_lc: int = 255) -> None:
+        self.apdus = apdus
+        self.max_lc = max_lc
+
+    @staticmethod
+    def _bytes(data: AnyStr) -> bytes:
+        if isinstance(data, (bytes, bytearray)):
+            return data
+        if isinstance(data, int):
+            return bytes([cast(int, data)])
+        if isinstance(data, str):
+            return bytes.fromhex(data)
+        raise TypeError(f"{data} cannot be converted to bytes")
+
+    @property
+    def apdus(self) -> Optional[ApduDict]:
+        return ApduSet._apdus if len(ApduSet._apdus.keys()) > 0 else None
+
+    @apdus.setter
+    def apdus(self, new_apdus: ApduDict, overwrite: bool = False) -> None:
+        """Sets a new CApsu internal dictionary if it wasn't set at instanciation time,
+        unless overwrite is True."""
+        if not self.apdus or overwrite is True:
+            if not isinstance(new_apdus, dict):
+                raise ValueError("Attribute newApdus must be a dictionary containing CApdu "
+                                 "instances as values")
+            ApduSet._apdus = new_apdus
```

Comment:
this setter seems overly complicated given how ApduSet is used.
I don't really see the point of going through an instance to then set a class variable. Can't self.apdus be a simple attribute that stores a dict?

Also I would advise against checking the type explicitly, it's not pythonic. As long as it quacks, it's a duck =]

Reply:
The idea behind this setter was to offer upper layers (i.e. the `DeviceAppProxy`-derived classes or even the tests) the possibility to dynamically change or update the set of APDUs used to communicate with the device, for instance if 2 apps are launched successively on a single device (e.g. for swap-like features maybe but I'm not even sure that 's how it works?). But actually I don't use that possibility and the setter is called only in `ApduSet`'s `__init__()` method to check the validity of the passed dict argument...

About the duck typing, sure, I can remove that type test, it would only result in an exception being raised later with a generic error message when the tests is started rather than when it is setup, making the issue a bit harder to pinpoint. 

Conclusion:
- Remove the setter & don't check type of `apdus` argument (framework user already has the type hints guiding him/her to passing the appropriate argument).

Reply:
👍